### PR TITLE
Asterisk (with chan_dongle) and FreePBX

### PR DESCRIPTION
**Pull request by m-anish**
This PR introduces the Asterisk and FreePBX applications for IIAB installs on top of Ubuntu 18.x based setups. 

~~Note: This is for testing only at this point. **Do not merge!**~~

You will have to atleast set pbx_install to True to have both apps install. And optionally set the enabled flag to enable them as well. 

Eagerly looking forwards to test results and polishing this further. We could potentially experiment with running this on debian 9.6 as well.

### CODE REVIEW

File: roles/pbx/tasks/asterisk.yml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    - /var/lib/asterisk
+    - /var/log/asterisk
+    - /var/spool/asterisk
+    - /usr/lib/asterisk
+
+- name: Asterisk - Set default user to asterisk in /etc/default/asterisk
+  lineinfile:
+      path: /etc/default/asterisk
+      regexp: 'AST_USER='
+      line: 'AST_USER="asterisk"'
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    line: 'AST_USER="asterisk"'
```

File: roles/pbx/tasks/asterisk.yml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    - /etc/asterisk
+    - /var/lib/asterisk
+    - /var/log/asterisk
+    - /var/spool/asterisk
+    - /usr/lib/asterisk
+
+- name: Asterisk - Set default user to asterisk in /etc/default/asterisk
+  lineinfile:
+      path: /etc/default/asterisk
+      regexp: 'AST_USER='
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    regexp: 'AST_USER='
```

File: roles/pbx/tasks/asterisk.yml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  with_items:
+    - /etc/asterisk
+    - /var/lib/asterisk
+    - /var/log/asterisk
+    - /var/spool/asterisk
+    - /usr/lib/asterisk
+
+- name: Asterisk - Set default user to asterisk in /etc/default/asterisk
+  lineinfile:
+      path: /etc/default/asterisk
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    path: /etc/default/asterisk
```

File: roles/pbx/tasks/asterisk.yml [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      regexp: 'AST_USER='
+      line: 'AST_USER="asterisk"'
+
+- name: Asterisk - Set default group to asterisk in /etc/default/asterisk
+  lineinfile:
+      path: /etc/default/asterisk
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    path: /etc/default/asterisk
```

File: roles/pbx/tasks/asterisk.yml [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      regexp: 'AST_GROUP='
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    regexp: 'AST_GROUP='
```

File: roles/pbx/tasks/asterisk.yml [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      line: 'AST_GROUP="asterisk"'
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    line: 'AST_GROUP="asterisk"'
```

File: roles/pbx/tasks/asterisk.yml [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+- name: Asterisk - Set default user to asterisk in /etc/asterisk/asterisk.conf
+  lineinfile:
+      path: /etc/asterisk/asterisk.conf
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    path: /etc/asterisk/asterisk.conf
```

File: roles/pbx/tasks/asterisk.yml [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      regexp: 'runuser ='
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    regexp: 'runuser ='
```

File: roles/pbx/tasks/asterisk.yml [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+      line: 'runuser = asterisk'
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    line: 'runuser = asterisk'
```

File: roles/pbx/tasks/asterisk.yml [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+- name: Asterisk - Set default group to asterisk in /etc/asterisk/asterisk.conf
+  lineinfile:
+      path: /etc/asterisk/asterisk.conf
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    path: /etc/asterisk/asterisk.conf
```

File: roles/pbx/tasks/asterisk.yml [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+      regexp: 'rungroup ='
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    regexp: 'rungroup ='
```

File: roles/pbx/tasks/asterisk.yml [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+      line: 'rungroup = asterisk'
```
Comment (holta) [inline roles/pbx/tasks/asterisk.yml]:
```suggestion
    line: 'rungroup = asterisk'
```

**Review by holta — COMMENTED** [review]
Ansible's general recommendation is 2-space indentation.

File: roles/pbx/tasks/asterisk.yml (hunk 1 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: sorry, vim wants to do this by default, and it is painful to fix it everytime. musthave sneaked thru

Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: will fix all these in one go

**Comment by m-anish** [comment]
Okay, some thoughts here... 

1. FreePBX wants node 10.x, so I am going to test it with that. This introduces a problem for us, as it makes it incompatible with sugarizer. 
2. The playbook has two parts. One to install Asterisk, and the other to install FreePBX. 
3. The Asterisk part is robust, and seems to install cleanly without creating issues. It will probably work on Debian Stretch, and perhaps even Buster, but I dont want to test that yet. 
4. The FreePBX part is more tricky. Typically, FreePBX wants to be the only thing it runs on the server. This either means we should do this out of a docker container, which I dont prefer for obvious performance reasons, or it means to tweak its standard config. Read on...
5. Pending: Create a custom freepbx.conf apache conf file, that serves freepbx through a virtualhost, in which the user is asterisk instead of www-data.

**Comment by m-anish** [comment]
Okay, freepbx 15 seems to install properly with nodejs 10.x. This has a few implications. 

1. Let us drop this for 6.7 - (if even we were remotely considering it) and target for 7.0
2. I am going to put a dependency on this. DO NOT install if sugarizer is going to be installed. 
3. Going to test this alongside nodered and mqtt for node 10.x as that should work _in theory_
4. So, this needs lots of testing because of the nodejs changes
5. I will make a separate PR that moves us to a nodesource based install generally. That seems to be the preferred way to install nodejs, whether 8.x or 10.x.

**Comment by m-anish** [comment]
Checklist:

- [x] Do NOT install if sugarizer is enabled
- [x] Do NOT install if nodejs version is not set to 10.x
- [x] Install Apache module to run as multiple users, since freepbx needs apache to run as user asterisk
- [x] Install Apache freepbx config file
- [x] Fix indentation issues
- [x] Create a bind9/dnsmasq host pbx.lan to serve freepbx. @jvonau @holta pls clarify how to go about this.
- [x] Only install freepbx once! Create and honour a flag for the same.
- [ ] Passes the testing checklist [here](https://github.com/iiab/iiab/pull/1405#issuecomment-455497996)
- [x] Speed up the playbook by running ansible install steps once (use 'creates' flags)
- [x] Add a README
- [x] Optionally, install chan_dongle with asterisk. 
- [x] Merge #1411 (or similar) prior to this. 
- [x] Disable installation and enabling by default on all: min, med, big installs

**Comment by m-anish** [comment]
Note: Test only after pulling in #1411 and setting nodejs_version to 10.x

File: roles/pbx/tasks/asterisk.yml (hunk 2 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

File: roles/pbx/tasks/asterisk.yml (hunk 3 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

File: roles/pbx/tasks/asterisk.yml (hunk 4 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

File: roles/pbx/tasks/asterisk.yml (hunk 5 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

File: roles/pbx/tasks/asterisk.yml (hunk 8 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

File: roles/pbx/tasks/asterisk.yml (hunk 6 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

File: roles/pbx/tasks/asterisk.yml (hunk 7 above)
Comment (m-anish) [reply roles/pbx/tasks/asterisk.yml]: 80b7d51 resolves this.

**Comment by m-anish** [comment]
**Okay. This is now good to test after locally merging #1411**

**Comment by m-anish** [comment]
Comprehensive test checklist:

- [x] Ensure that install does NOT run on any platform except ubuntu 18.x and 9.7
- [ ] Smoke test in gateway mode configuration, with pbx.lan properly set
- [x] Smoke test FreePBX on ubuntu 18.04 after successful install
- [x] Playbook should only install freepbx once and not fail if re-run
- [x] Install=False when Sugarizer=True or Nodejs=8.x
- [x] Install=True only when PBX=True and Nodejs=10.x
- [x] Test in a minimal install with `local_vars_min + PBX`
- [x] Test in a full install with `local_vars_big + PBX - Sugarizer`

**Comment by holta** [comment]
If this is well-tested & not disruptive to Sugarizer over the coming ~48h, we can consider this for inclusion in IIAB 6.7

**Comment by m-anish** [comment]
You can see from the checklist that I made good progress on the testing front, but since the scripts in asterisk and especially freepbx do many things, i would like atleast someone else other than me to also install and test, preferably from scratch, on bare metal. I would be much more confident then. Unfortunately, I dont have access to a NUC over the coming few days.

**Comment by m-anish** [comment]
Brought up to date with current master.

**Comment by holta** [comment]
@m-anish asks for testing on actual hardware (not just VM's).

CLARIF: Ubuntu Server 18.04 is required for now.

@jvonau suggesting 2 ways to deal to make pbx.lan work with dnsmasq:
1) /etc/hosts (use Ansible's lineinfile command? IIAB's network role likely not nec here?)
2) separate file sourced from dnsmasq.d

**Comment by m-anish** [comment]
Sure, I will add to my todo to edit this PR to edit /etc/hosts

**Comment by holta** [comment]
fyi Node.js 10.x is now part of IIAB (master branch iiab/iiab) if that helps Asterisk/FreePBX move forward:
https://github.com/iiab/iiab/blob/master/vars/default_vars.yml#L287-L289

**Comment by m-anish** [comment]
Very nice. I think I need to appropriately make changes to this playbook then.

File: roles/pbx/tasks/main.yml [hunk 13]
```diff
@@ -7,8 +7,8 @@
 
 - name: Install asterisk
   include_tasks: asterisk.yml
-  when: internet_available and is_ubuntu_18 and pbx_install and (not pbx_installed) and (not sugarizer_install) and (nodejs_version == "10.x")
+  when: internet_available and ((is_debian and ansible_distribution_major_version == "9") or is_ubuntu_18) and pbx_install and (not pbx_installed) and (nodejs_version == "10.x")
```
Comment (holta) [inline roles/pbx/tasks/main.yml]: @m-anish can you make this and Line 14 condition a bit more flexible, so Debian 10 and other Ubuntus (like 19.04 coming in April) and Raspbian-on-RPi might actually work?

Comment (m-anish) [reply roles/pbx/tasks/main.yml]: I will once I have tested those platforms. But want to stick to this for now.

Comment (holta) [reply roles/pbx/tasks/main.yml]: > * consider `is_debian and (not is_rpi)`

Another option might be `is_debian` (just like `is_ubuntu` ideally) simplifying in the future anyway!?

**Comment by holta** [comment]
@m-anish I'm going to go ahead and merge but I need your help testing/refining:
- `./runrole pbx` completes on an Ubuntu Server 18.04.1 VM (single network interface) but how do we test?
- Specifically how do I access http://pbx.lan/freepbx which does not work so far?
  - Running ./iiab-network did not help (pbx.lan still doesn't resolve)
  - Side Question: is "pbx pbx.lan" not required on Line 21 of [roles/network/tasks/hosts.yml](https://github.com/iiab/iiab/blob/master/roles/network/tasks/hosts.yml) ?  (Certainly it's on Line 13.)

**Comment by m-anish** [comment]
> * `./runrole pbx` completes on an Ubuntu Server 18.04.1 VM (single network interface) but how do we test?

On your host machine, you will need to add an entry to `/etc/hosts` like this one
`<VM's IP address>   pbx.lan`

And then open http://pbx.lan/freepbx to test.

**Comment by m-anish** [comment]
> Specifically how do I access http://pbx.lan/freepbx which does not work so far?
> 
>     * Running ./iiab-network did not help (pbx.lan still doesn't resolve)
> 
>     * Side Question: is "pbx pbx.lan" not required on Line 21 of [roles/network/tasks/hosts.yml](https://github.com/iiab/iiab/blob/master/roles/network/tasks/hosts.yml) ?  (Certainly it's on Line 13.)

Is this in gateway or appliance?

**Comment by holta** [comment]
> Is this in gateway or appliance?

Whatever's normal for a VM with a single network interface!

Log into it (192.168.0.161) if that's not clear enough?

**Comment by m-anish** [comment]
> > Is this in gateway or appliance?
> 
> Whatever's normal for a VM with a single network interface!
> 
> Log into it (192.168.0.161) if that's not clear enough?

VM would be appliance. Please follow the instructions [here](https://github.com/iiab/iiab/pull/1405#issuecomment-462098056)

**Pull request merged**